### PR TITLE
Fixing TriCore disasm instructions

### PR DESCRIPTION
**Pull request by bkoppelmann**
Hi,
while trying to integrate Capstone into QEMU, I came across some bugs.
Here are the fixes :)

Cheers,
Bastian

### CODE REVIEW

**Comment by XVilka** [comment]
@imbillow take a look please when you have time

**Comment by XVilka** [comment]
It should probably be fixed in TableGen files first, then regenerated though

File: arch/TriCore/TriCoreInstPrinter.c [hunk 1]
```diff
@@ -224,6 +224,19 @@ static void off4_fixup(MCInst *MI, uint64_t *off4)
...
 }
 
+static void off8_fixup(MCInst *MI, uint64_t *off8)
+{
+	switch (MCInst_getOpcode(MI)) {
+	case TRICORE_LD_A_sc:
+	case TRICORE_ST_A_sc:
+	case TRICORE_ST_W_sc:
+	case TRICORE_LD_W_sc: {
+		*off8 *= 4;
```
Comment (b1llow) [inline arch/TriCore/TriCoreInstPrinter.c]:
```suggestion
static void const8_fixup(MCInst *MI, uint64_t *const8)
{
	switch (MCInst_getOpcode(MI)) {
	case TRICORE_LD_A_sc:
	case TRICORE_ST_A_sc:
	case TRICORE_ST_W_sc:
	case TRICORE_LD_W_sc: {
		*const8 *= 4;
```

File: arch/TriCore/TriCoreInstPrinter.c [hunk 2]
```diff
@@ -235,6 +248,9 @@ static void print_zero_ext(MCInst *MI, int OpNum, SStream *O, unsigned n)
 		if (n == 4) {
 			off4_fixup(MI, &imm);
 		}
+		if (n == 8) {
+			off8_fixup(MI, &imm);
```
Comment (b1llow) [inline arch/TriCore/TriCoreInstPrinter.c]:
```suggestion
			const8_fixup(MI, &imm);
```

File: arch/TriCore/TriCoreDisassembler.c [hunk 3]
```diff
@@ -699,10 +756,24 @@ static DecodeStatus DecodeRRInstruction(MCInst *Inst, unsigned Insn,
...
-					     Decoder);
-		if (status != MCDisassembler_Success)
-			return status;
+		if (desc->OpInfo[0].OperandType == MCOI_OPERAND_REGISTER) {
+			switch (MCInst_getOpcode(Inst)) {
+			case TRICORE_ABSS_rr:
+			case TRICORE_ABSS_H_rr:
+			case TRICORE_ABS_H_rr:
+			case TRICORE_ABS_B_rr:
+			case TRICORE_ABS_rr: {
```
Comment (b1llow) [inline arch/TriCore/TriCoreDisassembler.c:765]: I checked and found that you still haven't fully fixed it here.
Can you please double-check the following instructions?

### `d, s2[, s1[, n]]`
ADDSC.A
ADDSC.AT
MOV.A
MOV.AA
MOV.D
CRC32.B
CRC32B.W
CRC32L.W

### `a`
UPDFL

**Review by b1llow — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
@bkoppelmann @imbillow do we have tests for instructions in question? If not - we should add in a testsuite.

File: arch/TriCore/TriCoreDisassembler.c (hunk 3 above)
Comment (bkoppelmann) [reply arch/TriCore/TriCoreDisassembler.c:765]: They have been correct before. Not sure why. These are the insns I sent to capstone as well as GNU objdump and they agree.
```
CODE = b'\x01\x23\x01\x16'  # addsc.a %a1,%a2,%d3,1
CODE += b'\x01\x34\x20\x26' # addsc.at %a2,%a3,%d4
CODE += b'\x01\x20\x30\x16' # mov.a %a1,%d2
CODE += b'\x01\x30\x00\x20' # mov.aa %a2,%a3
CODE += b'\x01\xb0\xc0\xa4' # mov.d %d10,%a11
CODE += b'\x4b\xbc\x60\xa0' # crc32.b %d10,%d11,%d12
CODE += b'\x4b\xbc\x30\xa0' # crc32 %d10,%d11,%d12
CODE += b'\x4b\xbc\x70\xa0' # crc32l.w %d10,%d11,%d12
CODE += b'\x4b\x0a\xc1\xa0' # updfl %d10
```

File: arch/TriCore/TriCoreInstPrinter.c (hunk 2 above)
Comment (bkoppelmann) [reply arch/TriCore/TriCoreInstPrinter.c]: Done

**Comment by bkoppelmann** [comment]
I can create tests. However, I'm not quite sure where to put them.  Is `suite/MC/TriCore/` the right place?

**Comment by XVilka** [comment]
>  Is `suite/MC/TriCore/` the right place?

Yes

**Comment by bkoppelmann** [comment]
> @bkoppelmann @imbillow do we have tests for instructions in question? If not - we should add in a testsuite.

Done

**Review by XVilka — APPROVED** [review]

File: arch/TriCore/TriCoreDisassembler.c (hunk 3 above)
Comment (b1llow) [reply arch/TriCore/TriCoreDisassembler.c:765]: I remembered that we can also change the order of operands in the td file.

**Comment by XVilka** [comment]
@imbillow take a look again please.

**Review by b1llow — APPROVED** [review]

**Comment by XVilka** [comment]
@kabeor this one is ready to merge. Also worth importing into v5 branch.

**Comment by kabeor** [comment]
Great! Thanks for your contribution!

**Pull request merged**